spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```
@@ -0,0 +1,268 @@
+    Broadcast<EncryptionManager> encryption = sparkContext.broadcast(encryptionManager);
+    boolean caseSensitive = Boolean.parseBoolean(spark.conf().get("spark.sql.caseSensitive", "false"));
+
+    RowDataRewriter rowDataRewriter = new RowDataRewriter(table, spec, caseSensitive, io, encryption,
+        rewriteDataFileSizeInBytes);
+
+    List<DataFile> addedDataFiles = rowDataRewriter.rewriteDataForTasks(taskRDD);
+    List<DataFile> currentDataFiles = filteredGroupedTasks.values().stream()
+        .flatMap(tasks -> tasks.stream().map(FileScanTask::file))
+        .collect(Collectors.toList());
+    replaceDataFiles(currentDataFiles, addedDataFiles);
+
+    return new RewriteDataFilesActionResult(currentDataFiles, addedDataFiles);
+  }
+
+  private Map<StructLikeWrapper, List<FileScanTask>> groupTasksByPartition(CloseableIterator<FileScanTask> tasksIter) {
+    Map<StructLikeWrapper, List<FileScanTask>> tasksGroupedByPartition = Maps.newHashMap();
+
+    try {
+      tasksIter.forEachRemaining(task -> {
+        StructLikeWrapper structLike = StructLikeWrapper.wrap(task.file().partition());
+        List<FileScanTask> taskList = tasksGroupedByPartition.getOrDefault(structLike, Lists.newArrayList());
+        taskList.add(task);
+        tasksGroupedByPartition.put(structLike, taskList);
+      });
+    } finally {
+      try {
+        tasksIter.close();
+      } catch (IOException ioe) {
+        LOG.warn("Faile to close task iterator", ioe);
+      }
+    }
+
+    return tasksGroupedByPartition;
+  }
+
+  private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles) {
+    boolean threw = true;
+
+    try {
+      RewriteFiles rewriteFiles = table.newRewrite();
+      rewriteFiles.rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
+      commit(rewriteFiles);
+      threw = false;
+
+    } catch (Throwable t) {
+      LOG.error("Failed to rewrite DataFiles", t);
```

This needs to re-throw the exception, not just swallow it.